### PR TITLE
[SPARK-37904][SQL] Improve RebalancePartitions in rules of Optimizer

### PR DESCRIPTION
**Pull request by ulysses-you**
### What changes were proposed in this pull request?

Improve `RebalancePartitions` in following rules:
- `NestedColumnAliasing`
- `CollapseRepartition `
- `EliminateSorts`
- `FoldablePropagation`
- `PropagateEmptyRelationBase`

### Why are the changes needed?

After SPARK-37267, we support do optimize rebalance partitions in everywhere of plan rather than limit to the root node. So It should make sense to also let `RebalancePartitions` work in all rules of Optimizer like `Repartition` and `RepartitionByExpression` did.

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

Add test in:
- `NestedColumnAliasingSuite`
- `CollapseRepartitionSuite`
- `EliminateSortsBeforeRepartitionSuite`
- `FoldablePropagationSuite`
- `PropagateEmptyRelationSuite`

### CODE REVIEW

**Comment by ulysses-you** [comment]
cc @cloud-fan @viirya

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreePatterns.scala [hunk 1]
```diff
@@ -111,6 +111,7 @@ object TreePattern extends Enumeration  {
   val PROJECT: Value = Value
   val RELATION_TIME_TRAVEL: Value = Value
   val REPARTITION_OPERATION: Value = Value
+  val REBALANCE: Value = Value
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreePatterns.scala:114]: nit: REBALANCE_PARTITION?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/PropagateEmptyRelationSuite.scala [hunk 2]
```diff
@@ -294,4 +294,20 @@ class PropagateEmptyRelationSuite extends PlanTest {
...
     comparePlans(optimized, expected)
   }
+
+  test("SPARK-37904: Improve rebalance in PropagateEmptyRelation") {
+    val emptyRelation = LocalRelation($"a".int)
+
+    // test root node
+    val plan1 = emptyRelation.rebalance($"a").analyze
+    val optimized1 = Optimize.execute(plan1)
+    val expected1 = emptyRelation.analyze
```
Comment (viirya) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/PropagateEmptyRelationSuite.scala:304]: nit: `expected1` and `expected2` are same. Maybe just `expected`.

**Review by viirya — COMMENTED** [review]
Looks good.

**Comment by HyukjinKwon** [comment]
cc @wangyum too

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreePatterns.scala (hunk 1 above)
Comment (ulysses-you) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreePatterns.scala]: changed to `REBALANCE_PARTITIONS`

**Review by viirya — APPROVED** [review]

**Review by wangyum — APPROVED** [review]

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by ulysses-you** [comment]
thank you all !

**Pull request closed without merging**